tests/activity_sticky_queues/activity_sticky_worker_workflow_test.py

```diff
@@ -0,0 +1,127 @@
+    return RETURNED_PATH
+
+
+@activity.defn(name="work_on_file_in_worker_filesystem")
+async def mock_work(path: str) -> str:
+    return CHECKSUM
+
+
+@activity.defn(name="work_on_file_in_worker_filesystem")
+async def mock_work_fail(path: str) -> str:
+    raise exceptions.ApplicationError("testing graceful failure")
+
+
+@activity.defn(name="clean_up_file_from_worker_filesystem")
+async def mock_cleanup(path: str) -> None:
+    pass
+
+
+async def test_processing_fails_gracefully(client: Client):
+    task_queue_name = str(uuid.uuid4())
+
+    @activity.defn(name="get_available_task_queue")
+    async def mock_task_queue():
+        return task_queue_name
+
+    async with Worker(
+        client,
+        task_queue=task_queue_name,
+        workflows=[tasks.FileProcessing],
+        activities=[
+            mock_download,
+            mock_work_fail,
+            tasks.clean_up_file_from_worker_filesystem,
+            mock_task_queue,
+        ],
+    ):
+        exception_catch = pytest.raises(WorkflowFailureError)
+        with exception_catch, mock.patch.object(tasks, "delete_file") as mock_del:
+            await client.execute_workflow(
+                tasks.FileProcessing.run,
+                id=str(uuid.uuid4()),
+                task_queue=task_queue_name,
+            )
+        mock_del.assert_called_once()
+
+
+async def test_worker_runs(client: Client):
+    task_queue_name_distribution = str(uuid.uuid4())
+    task_queue_workers = [str(uuid.uuid4()) for _ in range(3)]
+
+    @activity.defn(name="get_available_task_queue")
+    async def mock_task_queue():
+        return random.choice(task_queue_workers)
+
+    worker_distribution = Worker(
+        client,
+        task_queue=task_queue_name_distribution,
+        workflows=[tasks.FileProcessing],
+        activities=[
+            mock_task_queue,
+        ],
+    )
+    worker_1, worker_2, worker_3 = [
+        Worker(
+            client,
+            task_queue=queue,
+            activities=[
+                mock_download,
+                mock_work,
+                mock_cleanup,
+            ],
+        )
+        for queue in task_queue_workers
+    ]
+
+    # Should execute all fine
+    # p(pass| should fail) = 0.333 ^ (n_workers * n_iters)~ 6e-8
+    workflow_ids = [str(uuid.uuid4()) for _ in range(5)]
+    async with worker_distribution, worker_1, worker_2, worker_3:
+        for workflow_id in workflow_ids:
+            result = await client.execute_workflow(
+                tasks.FileProcessing.run,
+                id=workflow_id,
+                task_queue=task_queue_name_distribution,
+            )
+            assert result == CHECKSUM
+    # Check all events take place on the same random worker
+    workflow_executions = {
+        check_sticky_activity_count(
+            item.to_json_dict()["events"], task_queue_name_distribution
+        )
+        async for item in client.list_workflows().map_histories()
```

Comment:
Our unit tests run on the regular Temporal server and _also_ run with the time-skipping test server which does not implement list workflows (sorry!). You can change the top to:

```python
async def test_worker_runs(client: Client):
    if env.supports_time_skipping:
        pytest.skip("Test server doesn't support listing workflows")
```

And then don't forget the formatter. We're almost there, sorry for all of the CI checks :-)

Reply:
Nothing to apologise for! Better it breaks in CI than PROD